src/always-be-playing.js

```diff
@@ -135,6 +120,93 @@ export default class GapSkipper {
   }
 
   /**
+   * Handler for `waiting` events from the player
+   *
+   * @private
+   */
+  waiting_() {
+    let seekable = this.seekable();
+    let currentTime = this.tech_.currentTime();
+    let playlist = this.playlist();
+
+    if (this.tech_.seeking() || this.timer_ !== null) {
+      return;
+    }
+
+    if (this.checkFellOutOfLiveWindow_(playlist, seekable, currentTime)) {
+      return;
+    }
+
+    let buffered = this.tech_.buffered();
+    let nextRange = Ranges.findNextRange(buffered, currentTime);
+
+    if (this.checkVideoUnderflow_(nextRange, buffered, currentTime)) {
+      return;
+    }
+
+    this.checkGap_(nextRange, currentTime);
+  }
+
+  checkFellOutOfLiveWindow_(playlist, seekable, currentTime) {
+    if (!playlist ||
+        playlist.endList ||
+        !seekable.length ||
+        currentTime >= seekable.start(0)) {
+      return false;
+    }
+
+    let livePoint = seekable.end(seekable.length - 1);
+
+    this.logger_(`Fell out of live window at time ${currentTime}. Seeking to ` +
+                 `live point (seekable end) ${livePoint}`);
+    this.cancelTimer_();
+    this.tech_.setCurrentTime(livePoint);
```

Comment:
Design comment: I like the practice of splitting side-effecting operations from pure ones, all other things being equal. Instead of writing a function that checks-and-seeks-if necessary, you could write a check function and a seek-to-live function.